ports/atmel-samd/common-hal/busio/UART.c

```diff
@@ -246,6 +254,24 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
     usart_async_enable(usart_desc_p);
 }
 
+void common_hal_busio_uart_never_reset(busio_uart_obj_t *self) {
+#if (1)
+    never_reset_sercom(self->usart_desc.device.hw);
+    never_reset_pin_number(self->rx_pin);
+    never_reset_pin_number(self->tx_pin);
+    return;
```

Comment:
Please only include the active version.

Reply:
Good point!

I beleive it is what I mean, right?
```c
void common_hal_busio_uart_never_reset(busio_uart_obj_t *self) {
    for (size_t i = 0; i < MP_ARRAY_SIZE(sercom_insts); i++) {
        const Sercom *sercom = sercom_insts[i];
        Sercom *hw = (Sercom *)(self->usart_desc.device.hw);

        // Reserve pins for active UART only
        if (sercom == hw) {
            never_reset_sercom(hw);
            never_reset_pin_number(self->rx_pin);
            never_reset_pin_number(self->tx_pin);
        }
    }
    return;
}
```

Reply:
No? The if is true so the other one is used in your testing.

Reply:
I am not sure I understand your comment correctly.
Let me try to explain my idea.

* `common_hal_busio_uart_never_reset()` reserves the HW module (SERCOM) and correspending pins in order to avoid reseting of them it later on for some reason.
* `common_hal_busio_uart_never_reset()` is called, for instance, from `serial_early_init()` during initialization phase. The allocated UART object `debug_uart` is passed as argument value to `common_hal_busio_uart_never_reset()`;

### supervisor/shared/serial.c
```c
void serial_early_init(void) {
    #if defined(DEBUG_UART_TX) && defined(DEBUG_UART_RX)
    debug_uart.base.type = &busio_uart_type;

    const mcu_pin_obj_t *rx = MP_OBJ_TO_PTR(DEBUG_UART_RX);
    const mcu_pin_obj_t *tx = MP_OBJ_TO_PTR(DEBUG_UART_TX);

    common_hal_busio_uart_construct(&debug_uart, tx, rx, NULL, NULL, NULL,
        false, 115200, 8, UART_PARITY_NONE, 1, 1.0f, 64,
        buf_array, true);
    common_hal_busio_uart_never_reset(&debug_uart);
    #endif

    #ifdef NRF_DEBUG_PRINT
    _debug_uart_init();
    #endif
}
```

* `common_hal_busio_uart_never_reset()` function iterates over all available SERCOMs in the loop to check whether one of them matches SERCOM assigned to UART object `self` inside of `common_hal_busio_uart_construct()` before. The HW resources are reserved only if passed this condition is true. Did I miss anything?
### ports/atmel-samd/common-hal/busio/UART.c
```c
void common_hal_busio_uart_never_reset(busio_uart_obj_t *self) {
    for (size_t i = 0; i < MP_ARRAY_SIZE(sercom_insts); i++) {
        const Sercom *sercom = sercom_insts[i];
        Sercom *hw = (Sercom *)(self->usart_desc.device.hw);

        // Reserve pins for active UART only
        if (sercom == hw) {
            never_reset_sercom(hw);
            never_reset_pin_number(self->rx_pin);
            never_reset_pin_number(self->tx_pin);
        }
    }
    return;
}
```
----
BTW I took this code snippet from STM port.
### ports/stm/common-hal/busio/UART.c
```c
void common_hal_busio_uart_never_reset(busio_uart_obj_t *self) {
    for (size_t i = 0; i < MP_ARRAY_SIZE(mcu_uart_banks); i++) {
        if (mcu_uart_banks[i] == self->handle.Instance) {
            never_reset_uart[i] = true;
            never_reset_pin_number(self->tx->pin->port, self->tx->pin->number);
            never_reset_pin_number(self->rx->pin->port, self->rx->pin->number);
            break;
        }
    }
}
```

Reply:
I'm referring to the outer preprocessor if `#if (1)`. It is selecting between the two implementations and picking the first. Therefore, it's the version you tested with. I'm ok with you using the other version. You just need to remove the outer preprocessor `#if` and ensure that the second implementation works as expected.

Reply:
Oh, yes sure! I made the changes in the code.